test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+# Copyright (c) 2019 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.mininode import P2PInterface, P2PDataStore
+from test_framework.util import assert_equal, wait_until
+from test_framework.blocktools import create_block, create_coinbase
+from test_framework.messages import CTransaction, FromHex, msg_pong, msg_tx
+
+import time
+
+
+""" Test node eviction logic
+
+When the number of peers has reached the limit of maximum connections,
+the next connecting inbound peer will trigger the eviction mechanism.
+We cannot currently test the parts of the eviction logic that are based on
+address/netgroup since in the current framework, all peers are connecting from
+the same local address. See Issue #14210 for more info.
+
+Therefore, this test is limited to the remaining protection criteria:
+1.) Protect 4 nodes having sent us a block most recently
+2.) Protect 4 nodes having sent us a transaction most recently
+3.) Protect 8 nodes with the smallest minimum ping
+"""
+
+class SlowP2PDataStore(P2PDataStore):
+    def on_ping(self, message):
+        time.sleep(0.1)
```

Comment:
After #18638 this can use `setmocktime` instead of `sleep`?

Reply:
I don't think so. Logically this sleep should happen on the mininode side (of couple particular mininodes, not all).
Using mock time would sleep for *all* mininodes, but that doesn't satisfy us.

Unless the mininodes in the test are created in a very particular order, which would probably complicate things.

Reply:
Ah, thanks for the info!